### DISCUSSION TITLE
Transitive permissions on relations

### DISCUSSION THREAD

**Discussion by viters** [directus/directus#16208] [Feature Requests]
## Problem

Consider scenario with three collections, `posts`, `tags`, `posts_tags` with `read` permissions:
```js
// posts
{
  "_and": [
    { "status": { "_eq": "Published" } },
    { "created_at": { "_gte": "$NOW(-1 year)" } }
  ]
}

// tags
{ "status": { "_eq": "Published" } }
```

If you want to keep the API safe and clean, for `posts_tags` you need to set permissions:
```js
// posts_tags
{
  "_and": [
    { "posts_id": { "status": { "_eq": "Published" } } },
    { "posts_id": { "created_at": { "_gte": "$NOW(-1 year)" } } },
    { "tags_id": { "status": { "_eq": "Published" } } }
  ]
}
```

Notice that for each relation to `posts` you will need to duplicate the permissions for `posts`. This might require even more deep duplications if you have deeper relationships. It gets really messy with couple of rules to read an item or more roles with different rules.

I did not look into the implementation details yet, but I assume that if user requests `post` with `tags`, the `{ "created_at": { "_gte": "$NOW(-1 year)" } }` will need to be evaluated twice. It seems inefficient with larger data models.

## Solution

My idea is to allow inheritance of the permissions by checking if user can access the related item.

```js
// posts_tags
{
  "_and": [
    { "posts_id": { "_reachable": true } },
    { "tags_id": { "_reachable": true } }
  ]
}
```

That approach might allow caching of reachable objects for user, resulting in less permissions processing on relations, but most importantly it would be a lot less error prone on permissions management.

With structure `X <m2m> Y <m2m> Z` as `_reachable` is transitive, then:
1. (explicit) `Z` can be read if `Y` is `reachable`
2. (explicit) `Y` can be read if `X` is `reachable`
3. (implicit) `Z` can be read if `X` is `reachable`

**Comment by rijkvanzanten** [comment]
I think this is very similar (if not the same) to  https://github.com/directus/directus/discussions/3144 🙂

**Reply by viters** [reply]
I think your proposal is about an opposite operation. Although maybe they could be merged - I will think about it.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
